repos/repo_uri.go

```diff
@@ -0,0 +1,178 @@
+
+// Package repos defines a generic repository.
+package repos
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+
+	sce "github.com/ossf/scorecard/v2/errors"
+)
+
+var (
+	// ErrorUnsupportedhost indicates the repo's host is unsupported.
+	ErrorUnsupportedhost = errors.New("unsupported host")
+	// ErrorInvalidGithubURL indicates the repo's GitHub URL is not in the proper format.
+	ErrorInvalidGithubURL = errors.New("invalid GitHub repo URL")
+	// ErrorInvalidURL indicates the repo's full GitHub URL was not passed.
+	ErrorInvalidURL = errors.New("invalid repo flag")
+)
+
+// RepoURI represents the URI for a repo.
+type RepoURI struct {
+	repoType RepoType
+	url      repoURL
+	localDir repoLocalDir
+	metadata []string
+}
+
+type repoLocalDir struct {
+	path string
+}
+
+type repoURL struct {
+	host, owner, repo string
+}
+
+// RepoType is the type of a file.
+type RepoType int
+
+const (
+	// RepoTypeURL is for URLs.
+	RepoTypeURL RepoType = iota
+	// RepoTypeLocalDir is for source code in directories.
+	RepoTypeLocalDir
+)
+
+// NewFromURL creates a RepoURI from URL.
+func NewFromURL(url string) (*RepoURI, error) {
+	r := &RepoURI{
+		repoType: RepoTypeURL,
+	}
+
+	if err := r.Set(url); err != nil {
+		return nil, fmt.Errorf("%w", err)
+	}
+
+	return r, nil
+}
+
+// NewFromLocalDirectory creates a RepoURI as a local directory.
+func NewFromLocalDirectory(path string) *RepoURI {
+	return &RepoURI{
+		localDir: repoLocalDir{
+			path: path,
+		},
+		repoType: RepoTypeLocalDir,
+	}
+}
+
+func (r *RepoURI) SetMetadata(m []string) error {
+	r.metadata = m
+	return nil
+}
+
+func (r *RepoURI) SetURL(url string) error {
+	if err := r.Set(url); err != nil {
+		return fmt.Errorf("%w", err)
+	}
+	return nil
+}
+
+// Type method is needed so that this struct can be used as cmd flag.
+func (r *RepoURI) Type() string {
+	return "repo"
+}
+
+// GetType gives the type of URI.
+func (r *RepoURI) GetType() RepoType {
+	return r.repoType
+}
+
+// GetPath retusn the path for a local directory.
+func (r *RepoURI) GetPath() string {
+	return r.localDir.path
+}
+
+// GetURL returns a valid url for Repo struct.
+func (r *RepoURI) GetURL() string {
+	return fmt.Sprintf("%s/%s/%s", r.url.host, r.url.owner, r.url.repo)
+}
+
+// GetMetadata returns a valid url for Repo struct.
+func (r *RepoURI) GetMetadata() []string {
+	return r.metadata
+}
+
+// String returns a string representation of Repo struct.
+func (r *RepoURI) String() string {
+	return fmt.Sprintf("%s-%s-%s", r.url.host, r.url.owner, r.url.repo)
+}
+
+// Set parses a URI string into Repo struct.
+func (r *RepoURI) Set(s string) error {
+	const two = 2
+	const three = 3
+
+	// TODO: can we use this for pypi://, rubygems://, etc?
+	const httpsPrefix = "https://"
+	const filePrefix = "file://"
+
+	// Validate the URI and scheme.
+	if !strings.HasPrefix(s, filePrefix) &&
+		!strings.HasPrefix(s, httpsPrefix) {
+		return sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("invalid URI: %v", s))
+	}
+
```

Comment:
this only accepts `file://` and `https://`.
Anything else we want to support, like `org/name` or `github.com/org/name` can be done, but may be confusing for users who want to pass a file.
Thoughts?

Reply:
My only concern is backward compatibility. Our current users who might be used to `--repo=github.com/org/name` and --repo=`org/name` will now see failures. 

What if, anything starting with `file://` will be local repo else we continuing using the old model?

Reply:
Yeah, I agree. It could get confusing to users. I also like the approach of generalizing the URIs. Will wait for other also to chime in here, there are lot of interesting options to explore.

However, I'm hesitant about rolling this out a few days before making a big release. It seems a bit risky to me.